pkg/mic/mic.go

```diff
@@ -401,57 +441,166 @@ func (c *Client) makeAssignedIDs(azID *aadpodid.AzureIdentity, azBinding *aadpod
+	} else {
+		nodeMap[node.Name] = trackUserAssignedMSIIds{removeUserAssignedMSIIDs: []string{resourceID}, node: node}
+	}
+}
+
+func (c *Client) appendToAddListForNode(nodeMap map[string]trackUserAssignedMSIIds, resourceID string, node *corev1.Node) {
+	if trackList, ok := nodeMap[node.Name]; ok {
+		if trackList.addUserAssignedMSIIDs != nil {
+			trackList.addUserAssignedMSIIDs = append(trackList.addUserAssignedMSIIDs, resourceID)
+			trackList.node = node
+		} else {
+			trackList.addUserAssignedMSIIDs = []string{resourceID}
+			trackList.node = node
+		}
+		nodeMap[node.Name] = trackList
+	} else {
+		nodeMap[node.Name] = trackUserAssignedMSIIds{addUserAssignedMSIIDs: []string{resourceID}, node: node}
+	}
+}
 
-	if id.Spec.Type == aadpodid.UserAssignedMSI {
+func (c *Client) handleNodeErrors(nodesWithError []string, addList *[]aadpodid.AzureAssignedIdentity, deleteList *[]aadpodid.AzureAssignedIdentity, newAssignedIDs []aadpodid.AzureAssignedIdentity) error {
 
-		err = c.CloudClient.AssignUserMSI(id.Spec.ResourceID, node)
+	nodeIdentityList := make(map[string][]string)
+	for _, n := range nodesWithError {
+		// get the list of msi's currently on node
+		node, err := c.NodeClient.Get(n)
 		if err != nil {
-			glog.Error(err)
-			newErr := c.CRDClient.RemoveAssignedIdentity(assignedID)
-			if newErr != nil {
-				glog.Errorf("Error when removing assigned identity in create error path err: %v", newErr)
+			glog.Errorf("Lookup of node %s resulted in error %v", n, err)
+			continue
+		}
+		idList, err := c.getUserMSIListForNode(node)
+		if err != nil {
+			glog.Errorf("Getting list of msi's from node %s resulted in error %v", n, err)
+		}
+		nodeIdentityList[n] = idList
+	}
+
+	isNodeErrored := func(node string) bool {
+		for _, n := range nodesWithError {
+			if n == node {
+				return true
 			}
-			return err
 		}
+		return false
 	}
-	return nil
-}
 
-func (c *Client) removeAssignedIDsWithDeps(assignedID *aadpodid.AzureAssignedIdentity, inUse bool) error {
-	err := c.CRDClient.RemoveAssignedIdentity(assignedID)
-	if err != nil {
-		glog.Error(err)
-		return nil
+	if addList != nil && len(*addList) > 0 {
+		for _, createID := range *addList {
+			id := createID.Spec.AzureIdentityRef
+			binding := createID.Spec.AzureBindingRef
+			isUserAssignedMSI := c.checkIfUserAssignedMSI(id)
+
+			idExistsOnNode := c.checkIfMSIExistsOnNode(id, createID.Spec.NodeName, nodeIdentityList)
+			nodeHasErrors := isNodeErrored(createID.Spec.NodeName)
+
+			if !nodeHasErrors || idExistsOnNode {
+				// the identity was successfully assigned to the node
+				c.EventRecorder.Event(binding, corev1.EventTypeNormal, "binding applied",
+					fmt.Sprintf("Binding %s applied on node %s for pod %s", binding.Name, createID.Spec.NodeName, createID.Name))
+
+				glog.V(5).Infof("Initiating assigned id creation for pod - %s, binding - %s", createID.Spec.Pod, binding.Name)
+
+				if isUserAssignedMSI {
+					err := c.createAssignedIdentity(&createID)
```

Comment:
The assigned identity will only be created once the identity assignment on the node is successfully complete.